py-polars/polars/internals/dataframe/groupby.py

```diff
@@ -205,15 +205,21 @@ def apply(self, f: Callable[[pli.DataFrame], pli.DataFrame]) -> DF:
         ... )  # doctest: +IGNORE_RESULT
 
         """
-        by: Sequence[str]
+        by: list[str]
```

Comment:
`GroupBy.apply` only works on strings. This is a bit annoying to parse, but not trivial to solve.

Can't we somehow dispatch to the lazy groupby like the other methods?

Reply:
I wouldn't want to make `apply` good. It really shouldn't be used much.

Reply:
The reason I am asking, is that we can almost make a single `GroupBy` object for both eager and lazy, since most things are dispatched to lazy anyway.

This is blocking that optimization. Which is unfortunate, because as you say, it shouldn't even be used.

Reply:
I think it would need much more code on the lazy side, or cause slowdowns on the `apply` runtime.

Reply:
Yeah, I thought so. Better leave it like this for now.